samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -207,18 +190,21 @@ public void deleteStartpointForTask(SystemStreamPartition ssp, TaskName taskName
         TaskName taskName = taskModel.getTaskName();
         for (SystemStreamPartition ssp : taskModel.getSystemStreamPartitions()) {
           Startpoint startpoint = readStartpoint(ssp); // Read SSP-only key
-          if (Objects.nonNull(startpoint)) {
-            LOG.info("Grouping Startpoint keyed on SSP: {} to tasks determined by the job model.", ssp);
-            Startpoint startpointForTask = readStartpointForTask(ssp, taskName);
-            if (Objects.isNull(startpointForTask)) {
-              writeStartpointForTask(ssp, taskName, startpoint);
-              // Delete startpoint keyed by only this SSP
-              sspsToDelete.add(ssp);
-              LOG.info("Startpoint for SSP: {} remapped with task: {}.", ssp, taskName);
-            } else {
-              LOG.info("Startpoint for SSP: {} and task: {} already exists and will not be overwritten.", ssp, taskName);
-            }
+          if (startpoint == null) {
+            LOG.debug("No Startpoint for SSP: {} in task: {}", ssp, taskName);
+            continue;
           }
+
+          LOG.info("Grouping Startpoint keyed on SSP: {} to tasks determined by the job model.", ssp);
+          Startpoint startpointForTask = readStartpoint(ssp, taskName);
+          if (startpointForTask == null || startpointForTask.getCreationTimestamp() < startpoint.getCreationTimestamp()) {
+            writeStartpoint(ssp, taskName, startpoint);
+            sspsToDelete.add(ssp); // Mark for deletion
```

Comment:
Just to confirm: Is the idea here that if multiple taskNames map to a single SSP, then that SSP will just be put into this sspsToDelete set multiple times?

Reply:
That's correct and that's why I used a set instead of an arraylist.